pkg/ingester/instance.go

```diff
@@ -156,6 +162,15 @@ func (i *instance) getOrCreateStream(labels []client.LabelAdapter) (*stream, err
 	return stream, nil
 }
 
+// Return labels associated with given fingerprint. Used by fingerprint mapper. Must hold streamsMtx.
+func (i *instance) getLabelsFromFingerprint(fp model.Fingerprint) labels.Labels {
```

Comment:
you're not holding the streamMtx ? is that a problem ?

Reply:
we might not see the race because:
1) -race is off
2) the current unit test is synchronous.

I think you can improve the test below to run two push async and run manually the test with -race to verify we're good to go.

Reply:
I think I see why it is not a problem. Because line 89 takes the lock. 

I think it would still be useful to make the new test async (double push), so we can confirm everything is fine race wise.

Reply:
> I think I see why it is not a problem. Because line 89 takes the lock.

Exactly, line 89 and line 114. I'll make concurrent test. getLAbelsFromFingerprint is only used indirectly when calling `i.mapper.mapFP`, it's responsibility of caller of this method to hold the locks.

I will add the test and try to improve the documentation.

There is one more ugly place: mapper.go:81... it holds no lock inside the mapper, and relies on outside lock to update per-fingerprint map. This is directly from cortex, which uses locks per fingerprint. In Loki we use single global lock (streamMtx) so it is still fine, but not very clear.